### PR TITLE
Fix Progress gauge to reach 100%

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

This PR fixes the stuck progress bar when the rate flag is more than the system capabilities.
image

The most accurate representation was on nanoseconds so time.Duration.

### CODE REVIEW

File: gui/drawer.go [hunk 1]
```diff
@@ -125,17 +126,22 @@ L:
 	d.chartDrawing.Store(false)
 }
 
-func (d *drawer) redrawGauge(ctx context.Context, maxSize int) {
-	var count float64
-	size := float64(maxSize)
+func (d *drawer) redrawGauge(ctx context.Context, duration time.Duration) {
+	totalTime := float64(duration)
```
Comment (nakabonne) [inline gui/drawer.go:130]: I'd say we don't have to do redraw the gauge as soon as a result arrives anymore.

```suggestion
	ticker := time.NewTicker(redrawInterval)
	defer ticker.Stop()

	totalTime := float64(duration)
```

File: gui/drawer.go [hunk 2]
```diff
@@ -125,17 +126,22 @@ L:
...
 	d.widgets.progressGauge.Percent(0)
-	for {
+	for start := time.Now(); ; {
 		select {
 		case <-ctx.Done():
 			return
 		case <-d.gaugeCh:
```
Comment (nakabonne) [inline gui/drawer.go:136]: Let's remove the `gaugeCh` field from drawer.

```suggestion
		case <-ticker.C:
```

File: gui/drawer.go [hunk 3]
```diff
@@ -125,17 +126,22 @@ L:
...
-			count++
-			d.widgets.progressGauge.Percent(int(count / size * 100))
+			passed := float64(time.Since(start))
+			percent := int(passed / totalTime * 100)
+			// as time.Duration is the unit of nanoseconds
+			// small duration can exceed 100 on slow machines
+			if percent > 100 {
+				percent = 100
```
Comment (nakabonne) [inline gui/drawer.go:142]: We'd better avoid calling Termdash API more than needed.

```suggestion
				percent = 100
				continue
```

**Review by nakabonne — CHANGES_REQUESTED** [review]

**Comment by nakabonne** [comment]
@shubham14bajpai Thank you for your contribution! I've left some comments.

>sometimes the progress bar finishes milliseconds before the other sections of the GUI https://github.com/nakabonne/ali/issues/80#issuecomment-728129031

This is caused by entirely another problem with ali's drawing performance. If the ali is running at a high rate, the drawing may not be finished after the specified duration of time. This has been improved considerably, but there is still room for improvement.

**Comment by shubham14bajpai** [comment]
@nakabonne I have addressed review comments. PTAL.

**Review by nakabonne — APPROVED** [review]
Looks pretty neat. Thanks for your help!

**Pull request merged**